GameBoard.cpp

```diff
@@ -51,46 +59,104 @@ uint32_t GameBoard::xyToIndex(int x, int y) {
 
 }
 
-bool GameBoard::nextPointStatus(int x, int y) {
-	// TODO
-	return false;
-}
-
 uint8_t GameBoard::countNeighbors(uint32_t x, uint32_t y) {
 	// Would it be faster or slower to check if the value is 4 to avoid checking all pixels
 	// The if statements could lead to a slower processing time
 
 	// counts neighbors row by row
 	uint8_t count = 0;
 	uint32_t leftPoint = xyToIndex(x - 1, y - 1);
-	count += m_points.get(leftPoint);
-	count += m_points.get(leftPoint + 1);
-	count += m_points.get(leftPoint + 2);
+	count += m_cells.get(leftPoint);
+	count += m_cells.get(leftPoint + 1);
+	count += m_cells.get(leftPoint + 2);
 	leftPoint += c_sideLength;
-	count += m_points.get(leftPoint);
-	count += m_points.get(leftPoint + 2);
+	count += m_cells.get(leftPoint);
+	count += m_cells.get(leftPoint + 2);
 	leftPoint += c_sideLength;
-	count += m_points.get(leftPoint);
-	count += m_points.get(leftPoint + 1);
-	return count + m_points.get(leftPoint + 2);
+	count += m_cells.get(leftPoint);
+	count += m_cells.get(leftPoint + 1);
+	return count + m_cells.get(leftPoint + 2);
 }
 
 /*
 Chunk method definitions
 */
 
-GameBoard::Chunk::Chunk(GameBoard& gb, uint32_t offsetX, uint32_t offsetY) :
+GameBoard::Chunk::Chunk(GameBoard& gb, const Varient& varient, uint32_t offsetX, uint32_t offsetY) :
 	m_gb(gb),
 	c_offsetX(offsetX),
-	c_offsetY(offsetY) {}
+	c_offsetY(offsetY),
+	m_varient(varient),
+	m_border(BitArray(gb.c_chunkWidth * 2)) {
+	std::cout << "X: " << offsetX << " Y: " << offsetY << " test: " << m_varient.ySide << ' ' << m_varient.xSide << '\n';
+}
 
-// TODO: is this needed, or is Chunk::nextPointStatus sufficient
-uint8_t GameBoard::Chunk::countNeighbors(uint32_t x, uint32_t y) {
+bool GameBoard::Chunk::calcNextCellStatus(int x, int y) {
 	x += c_offsetX; y += c_offsetY;
-	return m_gb.countNeighbors(x, y);
+	bool alive = m_gb.getPoint(x, y);
+	uint8_t neighbors = m_gb.countNeighbors(x, y);
+
+	return neighbors == 3 || (alive && neighbors == 2);
 }
 
-bool GameBoard::Chunk::nextPointStatus(int x, int y) {
-	x += c_offsetX; y += c_offsetY;
-	return m_gb.nextPointStatus(x, y);
+void GameBoard::Chunk::processChunk() {
+	BitArray rows[2] = { BitArray(m_gb.c_chunkWidth), BitArray(m_gb.c_chunkWidth) };
+	bool currRow = true;
+	int y, x;
+	
+	for (x = 0; x < m_gb.c_chunkWidth; x++) {
+		rows[0].set(x, calcNextCellStatus(x, 0));
+	}
+
+	for (y = 1; y < m_gb.c_chunkWidth; y++) {
+		for (x = 0; x < m_gb.c_chunkWidth; x++) {
+			rows[currRow].set(y, calcNextCellStatus(x, y));
+		}
+		currRow = !currRow;
+		y--;
+		for (x = 0; x < m_gb.c_chunkWidth; x++) {
+			bool value = rows[currRow].get(x);
+			if (!addedToBorderBuffer(x, y, value)) {
+				m_gb.setPoint(x + c_offsetX, y + c_offsetY, value);
+			}
+		}
+		y++;
+	}
+	currRow = !currRow;
+	y--;
+	for (x = 0; x < m_gb.c_chunkWidth; x++) {
+		bool value = rows[currRow].get(x);
+		if (!addedToBorderBuffer(x, y, value)) {
+			m_gb.setPoint(x + c_offsetX, y + c_offsetY, value);
+		}
+	}

```

Comment:
```suggestion
        maxwh = m_gb.c_chunkWidth;    // compiler will take this out and it cleans up the code a little
	BitArray rowBuffers[2] = { BitArray(maxwh), BitArray(maxwh) };
	uint8_t hotBuffer = 0;
	int y, x;
	
	// put the first two rows into the buffer
	for (x = 0 ; x < maxwh ; x++) {
		rowBuffers[0].set(x, calcNextCellStatus(x, 0));
		rowBuffers[1].set(x, calcNextCellStatus(x, 1));
	}

	for (y = 0 ; y < maxwh ; y++) {
	        // write hot buffer the the board (y)
		for (x = 0 ; x < maxwh  ; x++) {
			bool value = rowBuffers[hotBuffer].get(x);
			if (!addedToBorderBuffer(x, y, value)) {
				m_gb.setPoint(x + c_offsetX, y + c_offsetY, value);
			}
		}
	        // write upcoming line into the buffer (y+2)
	        if ((y + 2) < maxwh) {
		        for (x = 0 ; x < maxwh ; x++) {
		    	        rowBuffers[hotBuffer].set(x, calcNextCellStatus(x, y+2));
		        }
		}
		// row y is written to the board, buffer now contains y+1 and y+2
		hotBuffer ^= 1;    // switch active buffer before next loop
	}
```

Reply:
I would not directly commit this. In fact I thiiiink this can still be improved. Line 111 in my suggestion could come out, and the "write upcoming line to buffer" step could be moved above the hot buffer write and changed to y+1.

Reply:
I didn't know the compiler removes function level renames. I would use that more so I dont have to write the path to a variable so often. I think the ordering makes more sense aswell, I knew duplicating the write row to board code twice probobly wasn't the best solution.